### PR TITLE
Implement User Authentication

### PR DESCRIPTION
**Pull request by Iamyuvraj143**
Manually setup for devise. User can sign up and login 
<kbd>home</kbd>

<kbd>signup</kbd>

<kpd>login</kbd>

<kpd>afterlogin</kbd>

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,14 @@
 class ApplicationController < ActionController::Base
+  before_action :set_current_user
+
+  def require_user_logged_in!
+    # allows only logged in user
+    redirect_to sign_in_path, alert: 'You must be signed in' if Current.user.nil?
+  end
```
Comment (kalpana19sharma) [inline app/controllers/application_controller.rb]: Do you think this should also be a private method?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,2 +1,14 @@
...
+
+  private
+  def set_current_user
```
Comment (kalpana19sharma) [inline app/controllers/application_controller.rb]:
```suggestion
  private

  def set_current_user
```

File: app/controllers/passwords_controller.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+class PasswordsController < ApplicationController
+  # allows only logged in users
+  before_action :require_user_logged_in!
+  def edit; end
+  def update
+    # update user password
```
Comment (kalpana19sharma) [inline app/controllers/passwords_controller.rb]: Use proper spacing between method definitions.
```suggestion
  before_action :require_user_logged_in!

  def edit; end

  def update
    # update user password
```

File: app/controllers/passwords_controller.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    if Current.user.update(password_params)
+      redirect_to root_path, notice: 'Password Updated'
+    else
+      render :edit
+    end
+  end
+    
+  private
+  def password_params
+    params.require(:user).permit(:password, :password_confirmation)
```
Comment (kalpana19sharma) [inline app/controllers/passwords_controller.rb]: Add an empty line above method definition.

File: app/controllers/passwords_controller.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  end
+  
+end
```
Comment (kalpana19sharma) [inline app/controllers/passwords_controller.rb]: Add new line at the end of the file and fix this issue in the whole PR

File: app/controllers/registrations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        # stores saved user id in a session
+        session[:user_id] = @user.id
+        redirect_to root_path, notice: 'Successfully created account'
+      else
+        render :new
+      end
+  end
+
+  private
+  def user_params
```
Comment (kalpana19sharma) [inline app/controllers/registrations_controller.rb:20]: Add an empty line above method definition.

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+class SessionsController < ApplicationController
+  
+  def new; end
+
+  def create
+    user = User.find_by(email: params[:email])
+    # finds existing user, checks to see if user can be authenticated
+    if user.present? && user.authenticate(params[:password])
```
Comment (kalpana19sharma) [inline app/controllers/sessions_controller.rb]:
```suggestion
    if user&.authenticate(params[:password])
```

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (Iamyuvraj143) [reply app/controllers/application_controller.rb]: ma'am In case if will use this method in other controller. That's why I haven't made it private !!

Comment (kalpana19sharma) [reply app/controllers/application_controller.rb]: So we can't access private methods in subclasses?

File: app/javascript/packs/application.js [hunk 8]
```diff
@@ -1,13 +0,0 @@
-// This file is automatically compiled by Webpack, along with any other files
```
Comment (kalpana19sharma) [inline app/javascript/packs/application.js]: Why did you remove this file?

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

**Review by kalpana19sharma — APPROVED** [review]
